Implement a virtual base class for plant component simulation

Myoldmopar opened the pull request:
# Yay

Restart the plant base class work again.  This time in small steps to see when any diffs are introduced.
### Fluidity
- This is **very** fluid right now, so don't base anything on this.  
### Phase 0
- Create a PlantComponent base class for `simulate()` only.
- Add a shared pointer on the plant data structure for later calling `simulate()` on an instance.
### Phase 1: Virtualize `SimXYZ()`

For each component to change, the following must be done:
- Make the component struct inherit the PlantComponent virtual base class
- Modify the component storage so that it is an array of shared pointers to objects, not objects themselves
- Create a static `factory` function in the component struct that verifies input has been "gotten" for that class of components, and returns the shared pointer (stored in the namespace array) to the component instance matching type/name, or executes a fatal error.
- Create a `simulate()` member function on the struct that either replaces the original `SimXYZ` namespace function, or wraps it nicely.
- Modify the `PlantManager` input to call the factory method for that particular type and assign the shared pointer on the plant structure.
- Modify the `PlantLoopEquip` file to call the member function instead of the `SimXYZ` function.
- Verify no-diffs or very explainable diffs.  But probably none.
### Phase 2

Begin virtualizing other parts of the components.  Likely possibilities are:
- `oneTimeInit()`
- `firstHVACInit()`

Once these are propagated through the entire set of plant components, I strongly expect a nice performance boost because we will eliminate so many checks littered heavily throughout the Sim and Init functions.
### Which Components?

The design must be exercised through a variety of component types.  The following list should suffice:
- [ ] Regular pipes (one inlet/outlet pair, very simple)
- [ ] Heat transfer pipes (same topology as regular pipes, but added complexity)
- [ ] Const COP chiller (up to 3 plant connections, lots of complexity)
- [ ] Pumps (one inlet/outlet, but different calling structure from other plant components)
### Tips during refactoring
- Watch for ambiguity with struct member variables and module level variables.  This is a huge possibility for a source of diffs and confusion.  Best approach: remove all module level variables first, or at least rename them something different from the struct member variable.  Once this version compiles fine, that should be much less of an issue moving forward.
- Analyze all static (non-member functions: Basically, those functions that don't have a CompIndex type of argument) to see if the functionality can be split out and called separately from the member function operations.
- Could consider doing all the member function work first to verify the worker functions can be objectified easily, then do Sim() after that.
### My To Do
- Go back and evaluate encapsulation, less public members and functions
- Go back and eliminate namespace level variables

Myoldmopar committed:
Implement trial base class approach for pipes

Myoldmopar committed:
Cleanup after reviewing diffs on GitHub

Myoldmopar committed:
Remove pipe heat transfer reporting struct (merge structs)

Myoldmopar committed:
Change simulate virtual function to include the initloopequip arg

Myoldmopar committed:
Intermediate commit; purposely renaming namespace variables to avoid ambiguity and verify no diffs so far

Myoldmopar committed:
Start forming pipe ht with some oo patterns, check for diffs

Myoldmopar committed:
Last commit before final simulate() restructuring for pipe ht

Myoldmopar committed:
Working version of pipe ht simulate refactor, check for diffs

Myoldmopar committed:
Add a basic Named class to move toward a unified VerifyName approach

Myoldmopar committed:
Update Named class and object storage to simplify and work with VerifyName

Myoldmopar committed:
Revert "Update Named class and object storage to simplify and work with VerifyName"

This reverts commit 17b038b228dd8eb7929b07128cbe7f8abd7f70ab.

Myoldmopar committed:
Reimplement changes for regular pipes; pipe HT still throwing diffs...

Myoldmopar committed:
Re-enable storing data in the PipeHT array, causes diffs

Myoldmopar committed:
AAAAHH @jasondegraw  a raw pointer; only for testing, dont worry

Myoldmopar committed:
Revert "AAAAHH @jasondegraw  a raw pointer; only for testing, dont worry"

This reverts commit e2116a8f357d59af51c80a91d606f2914f421fdf.

Myoldmopar committed:
Revert "Re-enable storing data in the PipeHT array, causes diffs"

This reverts commit 2df0a9f40c84a9238039b62fe5b04d30b1aca87b.

Myoldmopar committed:
Revert "Reimplement changes for regular pipes; pipe HT still throwing diffs..."

This reverts commit a8032a1cb81f1cb4b9365b5e68506cbc77e8faa9.

Myoldmopar commented:
OK, @mbadams5 you said you may have comments.  I'd like to get this merged in so I have a sturdy basis and can continue to flesh out our upcoming meeting.  Please add comments/concerns.  My plan is to get this merged in as-is with pipes and heat transfer pipes refactored, then do at least one chiller model completely before the meeting.

kbenne commented on src/EnergyPlus/DataPlant.hh:
```diff
@@ -597,6 +601,7 @@ namespace DataPlant {
 		int IndexInLoopSidePumps; // If I'm a pump, this tells my index in PL(:)%LS(:)%Pumps
 		Real64 TempDesCondIn;
 		Real64 TempDesEvapOut;
+		std::shared_ptr< PlantComponent > compPtr;
```
I'm going to take an unfiltered pass through this PR. Most of what I say will not be well thought out. I'm just blurting out my first impressions. 

shared_ptr ??? unique_ptr or something might be a better choice. Somebody should probably take responsibility for these objects. 

This also feels like a problem that is best solved up a level above Plant, but doing that might blow up your problem even more and in a way you are not ready to deal with.

Myoldmopar replied:
Yeah, I don't recall asking @kbenne for his thoughts...  

So I am trying to refactor in such a way that doesn't disrupt the existing framework (so that we can run to verify no-diffs at all stops), and so that the team is able to gradually move toward the refactored approach instead of a dead stop do-over.  Thus, I leave the objects to live in the namespace in a simple vector.  I tried leaving them as plain objects living in the namespace, not pointers at all, but well that went bad.  The higher level plant data structures need a pointer to the instance that lives at a particular location in the loop/branch/comp topology.  Since, for example, a chiller, can live at up to 3 locations, there are 3 variables in the loop/branch/comp topology that need to all point at the same instance of the chiller.  Thus it became a shared pointer.

kbenne replied:
Well hold on tight because here they come. my thoughts.

kbenne replied:
"I tried leaving them as plain objects living in the namespace, not pointers at all, but well that went bad." - yeah well the polymorphism wont work anyway with instances.

mbadams5 replied:
So I know people have an aversion to 'raw' pointers but what if they were just used for sharing of an object owned by something else? For example,

``` cpp
namespace EnergyPlus {
  namespace DataPlant {
    struct CompData
      {
        // Does not own PlantComponent object, just a reference basically
        PlantComponent * compPtr;
      }
  }
  namespace Pipes {
    // This owns the LocalPipeData instances for the lifetime of program 
    // since it is in a namespace.
    Array1D< LocalPipeData > LocalPipe;

    PlantComponent * LocalPipeData::factory( int objectType, std::string objectName ) {
      if ( GetPipeInputFlag ) {
        GetPipeInput();
        GetPipeInputFlag = false;
      }
      for ( auto * pipe : LocalPipe ) {
        if ( pipe->TypeOf == objectType && pipe->name == objectName ) {
          // returns pointer to object stored in Array. 
          // No need to worry about manual memory management since 
          // the Array owns that object and will dispose of it properly.
          return pipe;
        }
      }
      return nullptr;
    }
  }
  namespace PlantManager {
    void
    GetPlantInput()
    {
      // blah blah
      if ( SameString( this_comp_type, "Pipe:Adiabatic" ) ) { // This is awful by the way... sooo many string comparisons
        this_comp.TypeOf_Num = TypeOf_Pipe;
        this_comp.GeneralEquipType = GenEquipTypes_Pipe;
        this_comp.CurOpSchemeType = NoControlOpSchemeType;
        this_comp.compPtr = Pipes::LocalPipeData::factory( TypeOf_Pipe, CompNames( CompNum ) );
      }
      // blah blah
    }
  }
}
```

This is what you are trying to accomplish but without the overhead of shared_ptrs. Now this depends on the Array of objects to have a longer lifetime than any object that has member with a pointer to it.

I dunno, just a thought. Basically we need to figure out ownership and lifetime of objects.

:fire: :see_no_evil: :hear_no_evil: :speak_no_evil: :fire:

Myoldmopar replied:
I'm not against this.  I didn't plan on being the first one to put raw pointers in there, but I can be.  That would certainly simplify things.  It would allow us to keep GetInput the same since it is just building up an array of instances, storage of the instances would be the same, and we wouldn't have the overhead of the shared pointer.  I think this is good.

Any objections, concerns about proceeding with this @kbenne?  And @mjwitte if you have any concerns, feel free to chime in.

EnergyArchmage replied:
I think it is possible to come up with a design that avoids shared pointer.

On my new fan project, where I am doing a refactor of the old fans.cc, I have been having good success with holding the object instances in a std::vector of unique_ptr that sits in the namespace just outside the class.  Then when you need to store how to access this fan elsewhere, just store the integer vector index.  The namespace has a little get index routine from the fan name but everything else is in the class.  Changing to zero-based vector indexing is not too bad.  

}; //class FanSystem 

extern std::vector < std::unique_ptr <FanSystem> > fanObjs;

} // Fan namespace

you can check out HVACFan.hh and HVACFan.cc here if you want
https://github.com/EnergyArchmage/EplusFanDev

Myoldmopar replied:
@EnergyArchmage Thanks for this information.  I am always happy to see your successes and issues you have come across with your refactors.

I have a couple thoughts about your approach:
- I am hoping we move away "pretty much" fully from index-based operation, and instead move toward using all reference based (in my case a pointer) operations.  This will help us move away from things hanging globally altogether.
- In your approach, I'm not sure the value added in using unique pointers over storing just the instances themselves.  We can do object-oriented operations/structures with an array of instances, we could just be limited if we wanted to do polymorphism for different fan types, for example.

I think given how quick I am hoping to move while in Chicago, I am inclined to put this in the topic bucket for post-Chicago.  I think the solution that @mbadams5 described if sufficiently safe and relatively easy to understand without a bunch of homework, and also I don't feel like it paints us into a bad corner, we can use it as a stepping stone to get to something more formal.  And that "more formal" thing is likely something related to smarter pointers as you have shown here or in my initial efforts, but we'll see.

kbenne commented on src/EnergyPlus/Named.hh:
```diff
@@ -0,0 +1,75 @@
+
+namespace EnergyPlus {
+
+	class Named {
```
this appears to be your granddaddy base class. The name (pun intended or not) seems like a poor choice and painting you into a box. How about ... ModelObject :) .. just kidding. /ducks But maybe EnergyPlusObject or something.

mbadams5 replied:
I agree with @kbenne here. Is this only supposed to be used with objects that have a name field? What about objects that don't have a name field? Is name still going to be the unique identifier going forward? Anyways, I don't think "Named" is a particularly well 'named' class.

Another thought. You could do a setter/getter property instead of direct member variable access. What this gains is the ability to VerifyName (which could also come out of InputProcessor too) in the setter. The setter would need to take a container or iterator input parameter.

Myoldmopar replied:
These are good ideas.  However, I'm inclined to hold off on spending too much time refining this right now.  If you are super concerned about this base class painting us into a corner, I am happy to eliminate this base class altogether and just move the `std::string name` into the `PlantComponent` base class.  This would be simply moving the name field one level up instead of all the way to the granddaddy.  This would allow us to consider the inheritance chain later when I plan on doing that anyway.  What do you think of this solution?

mbadams5 replied:
@Myoldmopar If you get rid of this base class i'll need to change my `VerifyName()` implementation and more specifically, `FindItem()`. I use a `static_assert` to make sure at compile time that the passed in container elements are derived from `Named`, [seen here](https://github.com/NREL/EnergyPlus/pull/5426/files#diff-40acc796bb663e4257e28ed8f3684686R690).

I can not have this `static_assert` since (at this point) the only thing using this implementation are the new plant components. That is if you want to remove the Named base class.

kbenne commented on src/EnergyPlus/PipeHeatTransfer.hh:
```diff
@@ -99,30 +100,30 @@ namespace PipeHeatTransfer {
+	extern Real64 nsvEnvironmentTemp; // environmental temperature (surrounding pipe)
+	extern Real64 nsvEnvHeatLossRate; // heat loss rate from pipe to the environment
+	extern Real64 nsvFluidHeatLossRate; // overall heat loss from fluid to pipe
+	extern int nsvNumInnerTimeSteps; // the number of "inner" time steps for our model
```
Can these become members of your new class?

Myoldmopar replied:
They most definitely will eventually.  I was leaving them out to try to minimize the source diffs and still have no diff results.  I think devs will want to look at the diffs for each component before/after so they know what they are getting into with this refactoring, and moving those into the struct/class seemed to just muddy the diff.  Once this process is in full swing, it will be a relatively easy matter to move these into the class or eliminate them entirely.

kbenne commented on src/EnergyPlus/PipeHeatTransfer.hh:
```diff
@@ -311,116 +305,75 @@ namespace PipeHeatTransfer {
 
-	//==============================================================================
+		static
+		std::shared_ptr< PlantComponent > factory( int objectType, std::string objectName );
```
I'm being pedantic again, but i like the name create. Also not sure about that shared_ptr. Finally, maybe you could pass in the idf text or IdfObject or something. That way your factory doesn't have to access some global data to do the job.

mbadams5 replied:
This is more than just "create". It is more like "createOrFind".

I guess what is the advantage of storing things in an Array/vector if you just linearly search for the objects anyways? Why not use a `map` or `unordered_map` keyed on the name? The searching would be `O( logn )` and `O( 1 )` respectively AND they both would enforce unique names by default so no need to call VerifyName.

That said, both structures require more memory and can not be accessed by index. Although you seem to be moving away from that approach... Which I don't know if I agree with.

Myoldmopar replied:
Eventually I think the indeces will become fully moot, at least for these plant component and other "simple" component applications.  But right now there are still things that want to "grab" the data structures of these components by index.  We will _not_ be arguing that right now, eventually it will also go away, but for right now, leaving them in an Array1D that is index-base-1 keeps the possibilities for diffs and other stuff down to a minimum.

kbenne commented on src/EnergyPlus/PlantLoopEquip.cc:
```diff
@@ -308,19 +307,19 @@ namespace PlantLoopEquip {
 		//Pipe has no special types at the moment, so find it this way
 		if ( GeneralEquipType == GenEquipTypes_Pipe ) {
 			if ( EquipTypeNum == TypeOf_Pipe ) {
```
Isn't half of the point of doing this base class thing to introduce polymorphism so that you don't have these conditional switches? Can you not reduce this to one call to the base simulate?

Myoldmopar replied:
Yes, actually my hope is to eliminate this file entirely.  Again, this was somewhat delayed to show the devs the before and after of each individual component.  As things inherit PlantComponent, this will all diminish away...

jasondegraw replied:
Maybe collapsing all of the types that use `->simulate` into a single `if` clause would be "better":

```
if ( EquipTypeNum == TypeOf_Pipe  || EquipTypeNum == TypeOf_PipeSteam || EquipTypeNum == TypeOf_PipeExterior ...) {
```

mbadams5 replied:
I guess i'm confused why you need the conditionals as well. Why can't you just have one call:

`sim_component.compPtr->simulate( sim_component_location, FirstHVACIteration, InitLoopEquip );`

Myoldmopar replied:
Again, when all the components are transitioned into `PlantComponent`s, yes, this will fall down into a single line, eliminating the PlantLoopEquip huge blocks.  But for now, I planned on changing the lines 1-by-1 so that the other devs could see what I was doing easily.  I'd like to leave this one as-is for now.

kbenne commented on src/EnergyPlus/PlantManager.cc:
```diff
@@ -889,22 +887,27 @@ namespace PlantManager {
 							this_comp.TypeOf_Num = TypeOf_Pipe;
 							this_comp.GeneralEquipType = GenEquipTypes_Pipe;
 							this_comp.CurOpSchemeType = NoControlOpSchemeType;
+							this_comp.compPtr = Pipes::LocalPipeData::factory( TypeOf_Pipe, CompNames( CompNum ) );
```
Same comment here about using polymorphism.

Myoldmopar replied:
Somewhere inside the code, we need to have a switch block that checks the _type_ of object and interprets the idf data based on that type, right?  So even if we had a generic "factory" that you just passed in a bunch of idf text, it would have to create a chiller-specific data structure instance, for example, right?  This seems like the right place for it, given what I am trying to achieve at this point.  Looking beyond this initial refactoring I'm sure there are major other improvements, and we should certainly discuss them.  But I think this change should be suitable for making it better without painting us into a corner.

mbadams5 committed:
Fix VerifyName to work with container of shared_ptr

The new FindItem and VerifyName will work with containers of
shared_ptrs as well as regular objects. If the passed in iterator value
does not inherit from class Named then it will throw a compile time
assertion error.

mbadams5 commented on src/EnergyPlus/PlantComponent.hh:
```diff
@@ -0,0 +1,76 @@
+#define PLANTCOMPONENT_HH_INCLUDED
+
+#include <Named.hh>
+#include <PlantLocation.hh>
```
@Myoldmopar You can probably Forward Declare these instead of including the header file.

mbadams5 commented on src/EnergyPlus/Pipes.cc:
```diff
@@ -317,6 +237,9 @@ namespace Pipes {
 		NumLocalPipes = NumWaterPipes + NumSteamPipes;
 		LocalPipe.allocate( NumLocalPipes );
+		for ( int p = 1; p <= NumLocalPipes; ++p ) {
+			LocalPipe( p ) = std::shared_ptr< LocalPipeData >( new LocalPipeData() );
```
Should use `std::make_shared` instead. Or better yet, just use:

``` cpp
for ( int p = 1; p <= NumLocalPipes; ++p ) {
  LocalPipe.emplace_back( std::make_shared< LocalPipeData  >() );
}
```

Although i'm a little surprised `LocalPipe.allocate( NumLocalPipes );` doesn't default construct the objects...

Myoldmopar replied:
Welp, if we do the raw pointer storage on the Plant data structure and we just store instances in this array, this should be moot.

mbadams5 commented on src/EnergyPlus/Pipes.cc:
```diff
@@ -134,130 +135,49 @@ namespace Pipes {
-		//GET INPUT
+	std::shared_ptr< PlantComponent > LocalPipeData::factory( int objectType, std::string objectName ) {
+		// Process the input data for pipes if it hasn't been done already
 		if ( GetPipeInputFlag ) {
```
`GetPipeInputFlag` should really be part of the LocalPipeData class BUT since `LocalPipeData::factory` is static this doesn't work out well. `GetPipeInputFlag` would need to be a static member variable and we would run into the same issue when unit testing (and why there is clear_state now). Honestly, `LocalPipeData::factory` should pass in any data it needs to construct a new object. And that is a SINGLE object it creates, not all objects. 

What if the functionality is:

1) Try to find object in Array
2) If there, return pointer
3) If not, then process passed in input data or call to a (new) InputProcessor function that returns the input data for that specific object
4) Then construct said new object from input data. No grabbing from global data, only function calls.

No need for `GetPipeInputFlag` and `GetPipeInput()` is rolled into the factory basically. Makes unit testing easier since you can pass in or construct elsewhere the input data. Then call `LocalPipeData::factory` as many times as you want to construct or find that specific object.

Myoldmopar replied:
That sounds good.  But it seems like it is one step farther than I would like to take this initial pass.  I'm pushing to keep it like this for right now, and then proceed further with improvements as you described in a later refactoring.  The main reason is because when we try to scoot through _all_ the plant components, with the current way we don't have to modify GetInput...at all.  And that will really speed this process up.

mbadams5 commented on src/EnergyPlus/Pipes.hh:
```diff
@@ -59,12 +59,17 @@
 #include <EnergyPlus.hh>
 #include <DataGlobals.hh>
+#include <PlantComponent.hh>
+#include <PlantLocation.hh>
```
@Myoldmopar Again, can probably forward declare these.

mbadams5 committed:
Fix build error and use forward declarations

Myoldmopar commented:
OK, I'm really happy with the conversation here.  I think there are some great ideas that _will_ get done, but just maybe not in this first pass.  I want to make sure this first pass is suitable for sweeping through many components rather quickly.  I would like to make the change to store a pointer at the PlantLoop data structure level, and maybe rename the `factory` method so it makes more sense in its current formulation, and then merge this in.  After the March release, we can then look back at this and take some of the awesome ideas and make a plan for pushing them forward.

Is this suitable for wrapping this up prior to next week's meeting, @kbenne @mbadams5 ?

kbenne commented:
Well perfect is the enemy of good.

Myoldmopar committed:
Change storage back to instances and use a raw pointer on plant data structure

Myoldmopar committed:
Merge branch 'PlantCompBaseClassAgain' of github.com:NREL/EnergyPlus into PlantCompBaseClassAgain

Myoldmopar commented:
@kbenne @mbadams5 OK, so I was trying to minimize the discussion on this PR to just the relevant pieces needed to make Chicago productive.  I think most discussions have been deferred or addressed.  The only one that remains is the "Named" base class.  Quick backstory:
- During my work I used an array of shared pointers to objects in each namespace
- This broke the VerifyName call
- I added this Named base class to help get VerifyName working
- @mbadams5 then fixed the VerifyName function so that it would work
- Then we went back and stored actual instances in the namespace instead of shared pointers

So, I think at this point, I can rip out the Named class and just make sure all components still have their own `std::string Name` variable.  This is actually the least intrusive change for this PR and for devs making these changes at the meeting.  Do you have any issue with me backing the Named stuff back out?  If so, @mbadams5, is reverting the VerifyName just a matter of removing that `static_assert`?

If we can address this, I'd like to merge this in as-is, with no (numeric) diffs, so we have a nice basis for the meeting.

mbadams5 commented:
@Myoldmopar Yes, you can just comment out that static_assert. You will also need to remove the forward declarations. Also, if you use `Name` instead of `name` you will need to change the two `FindItem()` functions to use the upper case `Name`.

I do like the idea of the base class, I just don't think the name of the class is that great.

Myoldmopar committed:
Merge pull request #1 from energy-plus/PlantCompBaseClassAgain

Refactor Plant Components

Myoldmopar committed:
Cleanup simulate call, add static value virtuals, remove Named

Myoldmopar committed:
Merge pull request #2 from energy-plus/PrepDefCon3

Address some early issues

Myoldmopar committed:
Update vertical and slinky glhes

Myoldmopar committed:
Add CurLoad to PlantComponent API

Myoldmopar committed:
Add onInitLoopEquip virtual function to base class

Myoldmopar committed:
Merge branch 'develop' into LeeVerticalGLHE

Myoldmopar committed:
Update GLHEs for new API

Myoldmopar committed:
Add initLoopEquip call

Myoldmopar committed:
Merge pull request #3 from energy-plus/LeeVerticalGLHE

Vertical and Slinky GLHE

Myoldmopar committed:
Merge branch 'develop' of github.com:energy-plus/EnergyPlus into develop

mbadams5 committed:
Merge remote-tracking branch 'NREL/develop' into develop

mbadams5 committed:
Fix segfault and remove ambiguous calls

Myoldmopar committed:
Minor cleanups on PlantComponent

Myoldmopar committed:
Re-add UNUSED warning supression

Myoldmopar committed:
Add location to additional API functions

Myoldmopar committed:
Change CurLoad to pass byref in simulate()

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into PlantCompBaseClassAgain

Myoldmopar committed:
Revert the vertical/slinky transition out of develop

Myoldmopar committed:
Merge pull request #7 from energy-plus/RevertVerticalGHX

Revert the vertical/slinky transition out of develop

Myoldmopar committed:
Add override to simulate() for pipe and pipe ht

Myoldmopar committed:
quote FIX quote the diffs issue

Myoldmopar committed:
Merge remote-tracking branch 'hackathon-dev/develop' into PlantCompBaseClassAgain

Myoldmopar commented:
This PR is now obsolesced as the hackathon refactor changes will come in from energy-plus/develop...
